### ISSUE TITLE
Multi-Modelo con Grid Search (Agregación × Ventanas)

### ISSUE THREAD

**Issue by cuauhtemocbe** [cuauhtemocbe/btc-predictor#41]
## User Story

```
As a data scientist
I want to train a matrix of models combining different aggregation methods and time windows
In order to systematically find the optimal configuration that maximizes prediction accuracy
```

## Technical Context

**Current state:**
- `train_all_models()` exists (US-025) but trains all models with same configuration
- Trains 4 model types (Linear, LSTM, XGBoost, ARIMA) with fixed window_days=30
- Implementation: `workers/daily/trainer.py:351-526`

**Problem:**
- We can experiment with aggregation methods (issue #39) or windows (issue #40) separately
- But the optimal configuration might be a specific combination (e.g., "avg with 21 days")
- Manual experimentation is slow and error-prone

**Proposed solution:**
- Grid search over 2 dimensions: aggregation_method × window_days
- Train matrix of models: 4 methods × 4 windows = 16 configurations
- Automatically select best configuration based on validation error

**Research context:**
- Hyperparameter tuning is standard in ML (scikit-learn GridSearchCV)
- Systematic search prevents confirmation bias
- Validation set prevents overfitting to specific configuration

**Affected files:**
- `workers/daily/trainer.py` (new `train_grid_search()` function)
- `workers/daily/predictor.py` (multi-model prediction already exists)
- `api-service/templates/predictions.html` (dashboard to compare results)

**Dependencies:**
- Requires issues #39 and #40 completed (aggregation + windows implemented)
- Needs at least 90 days of historical data
- Requires sufficient compute resources (16 models training in sequence)

## Acceptance Criteria

```gherkin
Feature: Grid search over aggregation methods and time windows

  Background:
    Given the system has 90 days of daily BTC price data
    And aggregation methods are implemented: max, avg, median, weighted
    And time windows are supported: 7, 14, 21, 30 days

  Scenario: Configure grid search parameters
    Given I define aggregation_methods = ["max", "avg", "median", "weighted"]
    And I define window_days = [7, 14, 21, 30]
    When I call train_grid_search(aggregation_methods, window_days)
    Then it creates a configuration matrix with 16 combinations
    And each combination is defined as {"aggregation_method": X, "window_days": Y}

  Scenario: Train all grid search configurations
    Given grid search is configured with 4 methods × 4 windows = 16 configs
    When train_grid_search() executes
    Then 16 models are trained sequentially
    And each model has a unique name: "linear_{method}_{window}d_v1"
    And all models start with is_active=False

  Scenario: Store complete configuration in model params
    Given I train a model with config {"aggregation_method": "avg", "window_days": 21}
    When the model is saved to the database
    Then models.params contains {"aggregation_method": "avg", "window_days": 21, "validation_error_pct": X}
    And models.name is "linear_avg_21d_v1"

  Scenario: Calculate validation error for each configuration
    Given all 16 grid search models are trained
    And each model is evaluated on the same validation set (20% of data)
    When validation completes
    Then each model.params contains validation_error_pct
    And validation errors range from lowest to highest

  Scenario: Automatically activate best configuration
    Given grid search produced 16 models with validation errors
    When train_grid_search() completes
    Then the model with the lowest validation_error_pct is activated (is_active=True)
    And all other models remain inactive (is_active=False)
    And a log message reports the best configuration

  Scenario: Generate comparison report
    Given all 16 models are trained
    When I call generate_grid_search_report()
    Then it creates a markdown table with columns: method, window, validation_error_pct, training_time
    And the table is sorted by validation_error_pct ASC
    And the best configuration is highlighted

  Scenario: Multi-model prediction with grid search results
    Given I activate top 3 models from grid search
    When the predictor runs with --multi-model flag
    Then it generates 3 predictions for tomorrow
    And each prediction is associated with its model's configuration
    And the dashboard shows predictions grouped by configuration

  Scenario: Handle training failures gracefully
    Given 1 configuration (median, 7d) fails during training
    When train_grid_search() continues
    Then the remaining 15 models train successfully
    And the failed configuration is logged but doesn't stop the process
    And the best model is selected from successful trainings only

  Scenario Outline: Verify each configuration trains correctly
    Given aggregation_method=<method> and window_days=<window>
    When a model is trained with this configuration
    Then it fetches <window> days of data aggregated using <method>
    And the model name is "linear_<method>_<window>d_v1"

    Examples:
      | method   | window |
      | max      | 7      |
      | avg      | 14     |
      | median   | 21     |
      | weighted | 30     |
```

## Definition of Done

- [x] All Gherkin scenarios have passing automated tests
- [x] `train_grid_search()` function implemented in `workers/daily/trainer.py`
- [x] Grid search creates correct model names (e.g., "linear_avg_21d_v1")
- [x] Model.params stores both aggregation_method and window_days
- [x] Validation error calculated for all configurations
- [x] Best model automatically activated based on validation_error_pct
- [x] Grid search report generated (markdown table)
- [x] Dashboard displays multi-model predictions grouped by configuration
- [x] Training failures handled gracefully (don't stop grid search)
- [x] Test coverage >= 90%
- [x] Lint and tests green
- [x] Documentation updated in CLAUDE.md with grid search usage

## Effort: L (4-5 days)

**Breakdown:**
- Implement `train_grid_search()` function: 8 hours
- Generate comparison report: 4 hours
- Dashboard updates (grid search results): 6 hours
- Tests for grid search logic: 6 hours
- Run full grid search experiment (16 models): 4 hours
- Documentation and analysis: 4 hours

## Implementation Notes

**When to implement:**
- After issues #39 and #40 are completed and validated
- When we want to systematically optimize the system
- Requires compute resources (training 16 models may take 1-2 hours)

**Success metric:**
- Find optimal configuration that improves accuracy by at least 5% over baseline (max, 30d)
- Validate on hold-out test set (not used in training or validation)

**Grid search strategy:**
- Train all configurations with same train/validation split (fair comparison)
- Use 70% train, 20% validation, 10% hold-out test
- Store all models in database for future analysis
- Dashboard allows comparing any configuration

**Optimization (optional future work):**
- Parallel training (4 models in parallel) to reduce time
- Early stopping (skip remaining configs if one is clearly superior)
- Bayesian optimization instead of exhaustive grid search

**Reference:**
- Existing `train_all_models()` in `workers/daily/trainer.py:351-526`
- scikit-learn GridSearchCV pattern: https://scikit-learn.org/stable/modules/grid_search.html
- Multi-model prediction in `workers/daily/predictor.py:318-463`

**Example output:**
```
Grid Search Results (sorted by validation error):

| Method   | Window | Validation Error | Training Time |
|----------|--------|------------------|---------------|
| avg      | 21d    | 4.2%             | 12.3s         |
| weighted | 30d    | 4.5%             | 14.1s         |
| max      | 30d    | 4.8%             | 11.7s         | <- current baseline
| median   | 14d    | 5.1%             | 18.9s         |
| ...      | ...    | ...              | ...           |

✅ Best configuration: avg with 21 days (4.2% error)
   Improvement over baseline: 0.6 percentage points (12.5% relative improvement)
```
